### ISSUE TITLE
Bug in GC

### ISSUE THREAD

**Issue by knizhnik** [zenithdb/zenith#1047]
Sometimes GC deletes layers which are still needed.
Originally I faced with this problem when played on EC2 with large data size and pageserver is crashed because of disk space exhaustion.
My first idea was that it is caused by garbage collecting layers beyond disk consistent LSN:
https://github.com/zenithdb/zenith/pull/1004
But I failed to create test for it because deleted layers are restored by replaying WAL from safekeeper:
https://github.com/zenithdb/zenith/pull/1043
Configuration I have used on EC2 has not safekeepers.

But recently I was able to reproduce this problems locally without any restarts.
Just run read-only pgbench with scale 100 and 10 client  for a long time (1000 sec).
I got this errors:
```
pgbench: error: client 4 script 0 aborted in command 1 query 0: ERROR:  could not read block 13685 in rel 1663/13010/16404.0 from page server at lsn 1/ACE40AA8
DETAIL:  page server returned error: tried to request a page version that was garbage collected. requested at 1/ACE40AA8 gc cutoff 1/ACE4C878
progress: 770.0 s, 788.3 tps, lat 12.500 ms stddev 52.382
progress: 780.0 s, 730.6 tps, lat 12.206 ms stddev 32.596
pgbench: error: client 6 script 0 aborted in command 1 query 0: ERROR:  could not read block 134612 in rel 1663/13010/16396.0 from page server at lsn 1/B1CC5D48
DETAIL:  page server returned error: tried to request a page version that was garbage collected. requested at 1/B1CC5D48 gc cutoff 1/B1CCFB40
```
So something else is wrong in GC logic.

**Comment by knizhnik** [comment]
It may be not a problem at all: just current GC horizon (64Mb) is too small.
But I wonder if we need some more reliable mechanism for it.
Assume that some backend request some page from pageserver. Then vacuum or some other massive update transaction is activated and produce a lot of WAL. Amount of this WAL may be arbitrary large and when get_page_at_lsn request arrives at pageserver, this LSN is already behind GC horizon. So it is not clear which value of GC horizon can be considered as "safe".

**Comment by LizardWizzard** [comment]
There is one unfixed issue in gc that needs to be checked https://github.com/zenithdb/zenith/issues/707 There are some problems with reproduction so if you have some script that triggers the issue this would help too.
Also how many branches are involved in your test? Just one or more?

**Comment by LizardWizzard** [comment]
In my opinion GC shouldn't remove needed layers in any case. Even with GC horizon 0

**Comment by knizhnik** [comment]
> In my opinion GC shouldn't remove needed layers in any case. Even with GC horizon 0

Sorry, but it it is not possible. Even vanilla Postgres can report "snapshot too old" error.
And in our case lagged replica may request LSN which is too old (out of PITR interval).
The question is which value of GC horizon can be considered as "safe" for any workload.

**Comment by hlinnaka** [comment]
We chatted about this with @SomeoneToIgnore. My theory on this:

I think it's a race condition between `handle_get_page_at_lsn_request` and garbage collection. `handle_get_page_at_lsn_request` first checks what the latest record LSN is, by calling `wait_or_get_last_lsn`. Then it passes that LSN as argument to `get_page_at_lsn`. But it's possible that between those calls, more WAL is processed, advancing the latest record LSN, and garbage collection runs, advancing `latest_gc_cutoff_lsn` past the LSN that we calculated.

Would be good to have a reproducible test case for it first. For the fix, I think we need some mechanism to prevent GC from choosing a cutoff LSN value that is still being used in concurrent requests. With a lock of some sort I presume, although handle_get_page_at_lsn_request is a hot codepath, so we don't want to add much overhead to it with extra locking.

**Comment by knizhnik** [comment]
I wonder if this check is valid at all:
```
        ensure!(
            lsn >= latest_gc_cutoff_lsn,
            "tried to request a page version that was garbage collected. requested at {} gc cutoff {}",
            lsn, latest_gc_cutoff_lsn
        );
```

If page is not updated then it can have arbitrary small LSN.
And gc_cutoff is just last_lsn - gc_period.

So I think that this incorrect check should be removed.
Then race condition between handle_get_page_at_lsn_request and garbage collection is not so obvious.
Yes, user can request page which was already garbage collected. 
We should remove error in this case. 
I expect that layers lock should prevent race condition here: we either will not find required layer (because it was deleted by GC), 
either was able to extract page from this layer.


My patch ( #1128) address slightly different problem: when layers saved to the disk, but metadata files (and disk_consistent_lsn) are not yet updated. In this case GC can locate new image layer and remove old layers, but
if pageserver is restarted, ti removed all layers beyond  disk consistent LSN, and so some piece of table is disappeared.

**Comment by hlinnaka** [comment]
In summary, that check is correct. But we should have some kind of locking or something to prevent GC from choosing a cutoff that's newer than any "in-flight" GetPage requests. Perhaps add a barrier or something that allows GC to wait for any in-flight GetPage requests to be processed, before it advances latest_gc_cutoff_lsn.

**Comment by knizhnik** [comment]
> In summary, that check is correct. But we should have some kind of locking or something to prevent GC from choosing a cutoff that's newer than any "in-flight" GetPage requests. Perhaps add a barrier or something that allows GC to wait for any in-flight GetPage requests to be processed, before it advances latest_gc_cutoff_lsn.

I have stepped on the same rakes:(
Once I already understand why we can perform comparison with cutoff_lsn here, but then forgot.
But... If we think a little bit more about it...
Compute node is requesting last evicted LSN, or more precisely: largest LSN written to the disk. Assume that we produce a lot of WAL by updating small number of pages. So none of them is evicted. In this case last evicted LSN may be far in the past, may not it? In this case this check is still incorrect.

**Comment by hlinnaka** [comment]
> Compute node is requesting last evicted LSN, or more precisely: largest LSN written to the disk. Assume that we produce a lot of WAL by updating small number of pages. So none of them is evicted. In this case last evicted LSN may be far in the past, may not it? In this case this check is still incorrect.

To be precise, the compute node always requests the *latest* LSN, but *with a hint* of what the last evicted LSN was.

**Comment by hlinnaka** [comment]
So yes, the last evicted LSN can be very old. And that's OK. That's not what's passed to `get_page_at_lsn`.

**Comment by knizhnik** [comment]
O, now I inspected code of `wait_or_get_last_lsn` and understand why this check is correct.

**Comment by hlinnaka** [comment]
I was able to reproduce this readily with this recipe:

1. Apply this patch to add some delay to some GetPage requests:
```
diff --git a/pageserver/src/page_service.rs b/pageserver/src/page_service.rs
index 70ba7ec9..6b381734 100644
--- a/pageserver/src/page_service.rs
+++ b/pageserver/src/page_service.rs
@@ -484,6 +484,13 @@ impl PageServerHandler {
         let tag = RelishTag::Relation(req.rel);
         let lsn = Self::wait_or_get_last_lsn(timeline, req.lsn, req.latest)?;
 
+        // Add a 1s delay to some requests. The delayed causes the requests to
+        // hit the race condition from github issue #1047 more easily.
+        use rand::Rng;
+        if rand::thread_rng().gen::<u8>() < 5 {
+            std::thread::sleep(std::time::Duration::from_millis(1000));
+        }
+
         let page = timeline.get_page_at_lsn(tag, req.blkno, lsn)?;
 
         Ok(PagestreamBeMessage::GetPage(PagestreamGetPageResponse {
```

2. Run pgbench with
```
PGPORT=55432 PGUSER=zenith_admin pgbench -N -c10 -P1 -T120 postgres
```

3. At the same time with pgbench, in another terminal, force aggressive GC with:
```
/usr/bin/watch --interval=0 'psql postgres -p 64000 -h localhost -c "do_gc 308ddc0cd47c5e4c4a05744ddd180e1c 8905f92b2ee393ca9d29bf80d1788d1b 0"'
```

**Comment by knizhnik** [comment]
Still not able to reproduce it.
Did you mean the following patch:
```
diff --git a/pageserver/src/layered_repository.rs b/pageserver/src/layered_repository.rs
index 71a34d01..fa49c07e 100644
--- a/pageserver/src/layered_repository.rs
+++ b/pageserver/src/layered_repository.rs
@@ -1672,6 +1672,7 @@ impl LayeredTimeline {
     pub fn gc_timeline(&self, retain_lsns: Vec<Lsn>, cutoff: Lsn) -> Result<GcResult> {
         let now = Instant::now();
         let mut result: GcResult = Default::default();
+        let _checkpoint_cs = self.checkpoint_cs.lock().unwrap();
 
         let _enter = info_span!("garbage collection", timeline = %self.timelineid, tenant = %self.tenantid, cutoff = %cutoff).entered();
 
```

**Comment by hlinnaka** [comment]
> Still not able to reproduce it. Did you mean the following patch:

No, I meant what I posted.

I created a more complete python script to reproduce this at: https://github.com/zenithdb/zenith/tree/repro-issue-1047. I haven't been able to reproduce it without the patch I posted that adds a delay to some GetPage requests, the race condition is too tight. That patch is included in that branch, so if you run the `test_gc_aggressive.py` script that it adds, it fails.

**Comment by knizhnik** [comment]
> > Still not able to reproduce it. Did you mean the following patch:
> 
> No, I meant what I posted.
> 
> I created a more complete python script to reproduce this at: https://github.com/zenithdb/zenith/tree/repro-issue-1047. I haven't been able to reproduce it without the patch I posted that adds a delay to some GetPage requests, the race condition is too tight. That patch is included in that branch, so if you run the `test_gc_aggressive.py` script that it adds, it fails.

Sorry, there is definitely race condition between GC and checkpointer and without holding `checkpoint_cs` in GC I almost always get errors during test_write_amplification test. I have not proposed this fix as PR because I am still not sure if mutual exclusion of GC and checkpoint is the best solution (both can take arbitrary long time and both can have critical impact on performance and resource consumption).

With locking this mutex in checkpointer I was not able to reproduce the problem even with 0 GC horizon.
With your patch with random delay between wait_for_lsn and get_page_at_lsn I got
```
page server returned error: tried to request a page version that was garbage collected. 
```
error. But it seems to be expected in this case.

**Comment by knizhnik** [comment]
After thinking for a while about this problem, I realized, that it is not related with GC at all.
May be I missed something, but there seems to be the problem with layer_map:

Checkpointer creates image layer at inmem_layer.end_lsn. This layer contains pages which LSN may be less then end_lsn. But start_lsn of this layer is set to be equal to  inmem_layer.end_lsn. And this layer is registered in layer map with [start_lsn,start_lsn+1) interval. So any attempt to locate layer with smaller LSN will fail.

Assume that we have no GC at all. But we have aggressive checkpointer which very frequently creates new image layers.
But if get_page_at_lsn has deteriorated LSN, then we have to reconstruct page even through image of this page already reconstructed and stored in image layer.

So I see two directions of addressing this problem:
1. Postpone calculation of `latest_lsn`. Don't do it in wait_for_lsn, but in get_page_at_lsn. It may require changing `get_layer_for_read(seg,lsn)`, for example replacing it with `get_latest_layer_for_read(seg)`.
2. Change interval of image_layer from `[end_lsn,end_lsn+1)` to `[start_lsn,end_lsn+1)`. I am not sure what else has to be changed, how this layer will interfere with intersected delta layers,...

Before starting work in any of this direction, I want to get some confirmation from you. Do you have some other opinion about  the source of the problem and ways of its fixing?

**Comment by hlinnaka** [comment]
> With locking this mutex in checkpointer I was not able to reproduce the problem even with 0 GC horizon.
>With your patch with random delay between wait_for_lsn and get_page_at_lsn I got
> ```
> page server returned error: tried to request a page version that was garbage collected. 
> ```
> error. But it seems to be expected in this case.

Yeah, acquiring `checkpoint_cs` in GC only helps with this case happening from a checkpoint.

I think the simplest full fix would be to add yet another lock, for reads. The idea is that whenever you start a read operation at a particular LSN, you grab a new `read_lock` first. So `handle_get_page_at_lsn_request()` would do this:

1. Acquire `read_lock` in Share mode
2. Call wait_or_get_last_lsn() to get the current last record LSN
3. Call get_page_at_lsn() with that LSN
4. Release `read_lock`

Similarly, `evict_layer()` would first grab `read_lock`, then read last record LSN to perform the eviction and write the image layer at that LSN, and release `read_lock`.

GC would:
1. Read last record LSN
2. Acquire `read_lock` in Exclusive mode
3. Advance `latest_gc_cutoff_lsn`
4. Release `read_lock`
5. Perform GC

This way, GC cannot advance `latest_gc_cutoff_lsn` past an LSN that's still used by concurrent reads.

This adds some overhead to reads, as they need to acquire that lock. And when GC starts, incoming reads will need to wait for the in-progress reads to finish and for GC to grab the lock in exclusive mode. That couldt be pretty bad with checkpointer, as `evict_layer` creating the image layers can take a long time.

If we want to reduce that overhead, each page service handler thread could have their own read_lock, and GC would acquire each one separately. Or something like that. The point is for the GC to wait for any in-progress read operations to finish, before advancing `latest_gc_cutoff_lsn`.

**Comment by knizhnik** [comment]
Sorry, that I have been late with my comment (this morning I have three times changed my opinion about it: from not a problem at all, then fundamental bug in SMGR and to not related with GC:)
So why compute node may request not the latest LSN which page server have?
1. Read-only replica. Yes, it may be acceptable to return "version is already collected by GC" in this case if specified LSN is ut of PiTR interval. But is is good that we have to reconstruct page if there is existed image layer in this page.
2. Gap between wait_for_lsn and get_page_at_lsn. As I wrote we can try to somehow protect this gap with lock... Or change layer map to correctly deal with old LSNs...

**Comment by knizhnik** [comment]
Also this morning I once again realized that correctness of Zenith approach in SMGR is based on the assumption that Postgres never make some changes without holding  lock of underlying page and the page is marked as dirty before buffer is unpinned.
Looks like it is true (I have not found exception to this rule except initialization of gist/gin indexes). But in principle it is not required: we can first write WAL and only then update page. In this case our current implementation will not work correctly.

**Comment by hlinnaka** [comment]
> After thinking for a while about this problem, I realized, that it is not related with GC at all. May be I missed something, but there seems to be the problem with layer_map:
> 
> Checkpointer creates image layer at inmem_layer.end_lsn. This layer contains pages which LSN may be less then end_lsn. But start_lsn of this layer is set to be equal to inmem_layer.end_lsn. And this layer is registered in layer map with [start_lsn,start_lsn+1) interval.

Let's walk through this example and see if I understand it right. Let's imagine that at first layer_map contains just one image layer:

```
image: 0100
```

Some WAL is processed, and a checkpoint happens. A new delta and a new image layer is created:

```
image: 0100
delta: 0100-0200
image: 0200
```

> So any attempt to locate layer with smaller LSN will fail.

No. In the above example, if you request a page with e.g LSN 0150, it will find the last image layer before that LSN at 0100, and apply any WAL records needed for that page from the delta layer 0100-0200 to reconstruct the page as of LSN 0150.

Only GC removes the old layers. Until GC happens, they're still accessible.

> Assume that we have no GC at all. But we have aggressive checkpointer which very frequently creates new image layers. But if get_page_at_lsn has deteriorated LSN, then we have to reconstruct page even through image of this page already reconstructed and stored in image layer.

I didn't understand this paragraph. What's the problem?

> So I see two directions of addressing this problem:
> 
>     1. Postpone calculation of `latest_lsn`. Don't do it in wait_for_lsn, but in get_page_at_lsn. It may require changing `get_layer_for_read(seg,lsn)`, for example replacing it with `get_latest_layer_for_read(seg)`.

Yeah, we could explicitly pass the information that we're looking for the latest LSN. But you could still have the same race condition within the get_page_at_lsn(). We'd still need some kind of an interlock to prevent GC from removing the page version we're about to read. Maybe we could make `get_page_at_lsn()` tolerant of that, by adding more retries if a layer gets removed just when we're about to step into it or something. But I think it would be harder with the checkpointer, it certainly needs to be able to read a consistent snapshot of a segment at a particular LSN, when it creates a new image layer. Also, would be nice to use the same interlock mechanism with read-only replicas that are not requesting the latest LSN.

>     2. Change interval of image_layer from `[end_lsn,end_lsn+1)` to `[start_lsn,end_lsn+1)`. I am not sure what else has to be changed, how this layer will interfere with intersected delta layers,...

Yeah, I think that's a non-starter. It would make it a lot harder to reason about what happens.

**Comment by knizhnik** [comment]
> Only GC removes the old layers. Until GC happens, they're still accessible.
Yes, it is true.
Assume that there is no GC and we run the same example with random sleep between wait_for_lsn and get_page_at_lsn.
It may happen that master compute node will request LSN 0150 (with latest=true) flag and wait_for_lsn returns 150.
But before get_page_at_lsn is executed, `image: 0200` is created. But get_page_at_lsn  will still reconstruct requested page from `delta: 0100-0200` instead of taking ready page from  `image: 0200`.

It is not a bug, just inefficiency. But it seems to me that it illustrate that the source of the problem is not in GC.

**Comment by hlinnaka** [comment]
> Sorry, that I have been late with my comment (this morning I have three times changed my opinion about it: from not a problem at all, then fundamental bug in SMGR and to not related with GC:) 

:-D. Yeah, this stuff is complicated.

> So why compute node may request not the latest LSN which page server have?
> 
> 1. Read-only replica. Yes, it may be acceptable to return "version is already collected by GC" in this case if specified LSN is ut of PiTR interval. But is is good that we have to reconstruct page if there is existed image layer in this page.

Right. And we should have a mechanism that prevents GC from removing page versions that are still needed by read-only replicas. (We don't currently.) But if it nevertheless happens, we should reliably detect that case in the page server and give a decent error message like that.

> 2. Gap between wait_for_lsn and get_page_at_lsn. As I wrote we can try to somehow protect this gap with lock... Or change layer map to correctly deal with old LSNs...

The layer map is correct AFAICS. Let's add a lock to fix the gap.

> Also this morning I once again realized that correctness of Zenith approach in SMGR is based on the assumption that Postgres never make some changes without holding lock of underlying page and the page is marked as dirty before buffer is unpinned. Looks like it is true (I have not found exception to this rule except initialization of gist/gin indexes). But in principle it is not required: we can first write WAL and only then update page. In this case our current implementation will not work correctly.

Hmm, wouldn't that break things in vanilla PostgreSQL too? If you write WAL and release the lock, checkpoint could run before you actually modify the page. If you then crash and restart, the WAL record would not be replayed, so the update would be lost. So I think this is a safe assumption.

**Comment by knizhnik** [comment]
> Hmm, wouldn't that break things in vanilla PostgreSQL too? If you write WAL and release the lock, checkpoint could run before you actually modify the page. If you then crash and restart, the WAL record would not be replayed, so the update would be lost. So I think this is a safe assumption.

But it happens within one transaction. So if Posgres is crashed, then everything changed by this transaction should be undone. So it is correct not to replay this part of WAL, isn't it?

**Comment by hlinnaka** [comment]
On 17/01/2022 11:36, Konstantin Knizhnik wrote:
>     Hmm, wouldn't that break things in vanilla PostgreSQL too? If you
>     write WAL and release the lock, checkpoint could run before you
>     actually modify the page. If you then crash and restart, the WAL
>     record would not be replayed, so the update would be lost. So I
>     think this is a safe assumption.
> 
> But it happens within one transaction. So if Posgres is crashed, then 
> everything changed by this transaction should be undone. So it is 
> correct not to replay this part of WAL, isn't it?

Depends on what the update is. Not all updates are transactional, and 
even for transactional updates, it depends. E.g. for INSERT, if the 
index tuple already has been inserted, the corresponding heap tuple is 
present too, even if it the transaction aborts. This is all 
hypothetical, INSERT performs the locking correctly of course, but it's 
not safe in general that you can just skip the WAL records that were 
inserted by an aborted transaction.

- Heikki

**Comment by knizhnik** [comment]
The fact that page is 

> On 17/01/2022 11:36, Konstantin Knizhnik wrote: Hmm, wouldn't that break things in vanilla PostgreSQL too? If you write WAL and release the lock, checkpoint could run before you actually modify the page. If you then crash and restart, the WAL record would not be replayed, so the update would be lost. So I think this is a safe assumption. But it happens within one transaction. So if Posgres is crashed, then everything changed by this transaction should be undone. So it is correct not to replay this part of WAL, isn't it?
> Depends on what the update is. Not all updates are transactional, and even for transactional updates, it depends. E.g. for INSERT, if the index tuple already has been inserted, the corresponding heap tuple is present too, even if it the transaction aborts. This is all hypothetical, INSERT performs the locking correctly of course, but it's not safe in general that you can just skip the WAL records that were inserted by an aborted transaction. - Heikki

Ok, as far as there are no such places in Postgres tight now (I hope:),
then we should not try to fix unexisted problem (main principle of extreme programming:)

But in principle, if some update not yet changed somedatabase pages, then correspondent WAL records can be skipped during recovery. If it happen during page creation, then it can not interfere with other transaction so locking is not needed. But fortunately optimization scenario in this case is usually opposite: first write pages and then fix them in WAL.

**Comment by hlinnaka** [comment]
I believe this issue has been fixed by PR #1139, closing.

**Comment by knizhnik** [comment]
Yes, it was fixed by [f70a5cad61e756f3185522ff14daf0de9a117ec8]
